design-proposals/feature-lifecycle.md

```diff
@@ -0,0 +1,324 @@
+  a newly-proposed ("Alpha") feature in a controlled environment, to see
+  if it makes sense to me.
+- As a KubeVirt cluster operator, I would like to evaluate an
+  undergraduate ("Beta") feature in a real-life environment with actual users.
+- As a KubeVirt cluster operator, I would like to keep using a feature
+  that got graduated after I used it during the "Beta" evaluation period.
+- As a KubeVirt cluster operator, I would like to know that
+  an experimental ("Alpha") feature is planned to be removed.
+- As a KubeVirt cluster operator, I would like to know that
+  an undergraduate ("Beta") feature is planned to be removed.
+- As a KubeVirt cluster operator, I would like to stop using a feature
+  that got removed.
+
+### Repos
+This is a cross repo project policy under the
+[kubevirt](https://github.com/kubevirt) organization.
+
+## Proposal Design
+The proposal on how to define a feature lifecycle is influenced by
+processes and policies from the Kubernetes project.
+These sources are scattered around, each focusing on different
+aspects of a feature:
+- [Feature Gates](https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/)
+- [Graduation](https://kubernetes.io/blog/2020/08/21/moving-forward-from-beta/)
+- [Changing the API](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md)
+- [Deprecation](https://kubernetes.io/docs/reference/using-api/deprecation-policy/)
+
+The proposal takes the top-down approach, starting with the high level
+flow that a common feature will traverse through.
+Continuing with actions that need to be taken and timeline suggestions.
+
+Both feature graduation and discontinuation flows are covered.
+Including the implications on users.
+
+Depending on individual topics, follow-up proposals may extend the basic
+points raised in this proposal.
+
+### Terminology
+Feature Gates and feature configuration are often used interchangeable or not differentiated.
+- Feature Gate: A flag that controls the presence or availability of a feature in the cluster.
+- Feature configuration: Cluster or workload level configuration that allows an admin or user
+  (depending on the feature) to control aspects of a feature operation.
+  A common usage is to determine if features are opt-in or opt-out by default.
+
+### Feature Stages
+A feature is expected to pass in the following order through the following stages:
+1. Enhancement proposal.
+2. Implementation.
+3. Release as Alpha (experimental).
+4. Release as Beta (pre-release for evaluation).
+5. Release as General Availability (graduation).
+6. Removal.
+
+Starting from the Alpha release, it can be removed with restrictions that
+depend on the release stage (Alpha, Beta, GA).
+
+[Removal](#removal) of features is widely discussed later
+in this proposal.
+
+#### Enhancement proposal
+As the first step for introducing a new feature, a formal proposal is
```

Comment:
> I think that this is inherently a gray area.

The definition of what a `feature` represents is indeed "gray".

We could invest and define it well, but I feel it is an area of potential debate that will "hijack" the core proposal to a side discussion and increase the time to get this in.

Can we consider the term `feature` a black box, which one could debate about outside this proposal? Or perhaps, to open a follow up, even on this same file, with an attempt to define it well?

Just not to leave my response too general to your examples, I will try to respond to it:

> For example, let's say that a developer introducing a new feature goes through the phases you presented above (proposal, alpha, beta, GA). Then, after the feature is GAed, the developer wants to extend the API by presenting a new field. Would you say that in this case the developer needs go through the phases again, needing a new proposal, etc?

Between proposal and Beta (included), it is surely not a new feature.
Once it got GA, extending the feature with new API is surely a new feature, but it could be extended through the existing proposal document. Should we require a new fresh proposal/document? I am not sure, I tend to say `yes` but I would go with how things are followed in other similar projects (Kubernetes, Openshift).

> I'd also add that if there's an API change, someone from SIG-API would need to examine it, possibly tag more people for discussion, and finally approve it if it's valid.

I would prefer these technicalities to be treated in follow up proposals or guidelines.
There have been several points that I intentionally tried to treat without providing the implementation details, just to allow the "what" to be agreed on and leave the "how" for a later discussion.
I find this a more practical phased approach to get things in, because the how will surely have a lot of alternatives and agreement may be even harder there.

Reply:
> Can we consider the term feature a black box, which one could debate about outside this proposal? Or perhaps, to open a follow up, even on this same file, with an attempt to define it well?

That's completely fine by me as long as we understand that since the policy is not complete yet we would probably need to re-think if a DP were needed on a case-to-case basis. That might even be a good thing, since we'll probably be debating on how to apply this proposal and by the time we'll get to a follow-up we might get there with more experience.

I think that this PR is a great case study: https://github.com/kubevirt/kubevirt/pull/11500.

I can see how we get into debates over such PRs, and that's not a bad thing, but on the other hand we have to treat such cases as something to discuss and not as cases that necessarily demand a DP since we haven't finalized the policy and can't precisely define a feature yet.

Reply:
I opened an issue about defining a feature.
https://github.com/kubevirt/community/issues/285

Please confirm that this thread can be closed.